### PR TITLE
Add utility classes for new statistics

### PR DESCRIPTION
**Pull request by nafur**
This PR introduces two new sets of classes used for the new statistics setup.
The first set are the statistic values that hold the actual data and will live in the new statistics registry itself.
The second set are proxy objects: they only hold a pointer to the value classes and implement all the modifiers.

The code is not used yet, but replaces the code in the `stats_*` files in a subsequent PR.

### CODE REVIEW

**Comment by nafur** [comment]
Note that the `statistics_stats.*` are not yet registered in cmake. As the classes have the same name as the old ones, this would result in linker errors...

File: src/base/configuration.h [hunk 1]
```diff
@@ -48,7 +48,13 @@ class CVC4_EXPORT Configuration
 
   static bool isDebugBuild();
 
-  static bool isStatisticsBuild();
+  static constexpr bool isStatisticsBuild() {
```
Comment (aniemetz) [inline src/base/configuration.h:51]: Format

File: src/util/statistics_stats.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+IntStat& IntStat::operator=(int64_t val)
+{
+  if constexpr (Configuration::isStatisticsBuild())
+  {
+    d_data->d_value = val;
+  }
+  return *this;
+}
```
Comment (aniemetz) [inline src/util/statistics_stats.cpp:41]:
```suggestion
}

```

File: src/util/statistics_stats.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+IntStat& IntStat::operator++()
+{
+  if constexpr (Configuration::isStatisticsBuild())
+  {
+    d_data->d_value++;
+  }
+  return *this;
+}
```
Comment (aniemetz) [inline src/util/statistics_stats.cpp:49]:
```suggestion
}

```

File: src/util/statistics_stats.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+IntStat& IntStat::operator++(int)
+{
+  if constexpr (Configuration::isStatisticsBuild())
+  {
+    d_data->d_value++;
+  }
+  return *this;
+}
```
Comment (aniemetz) [inline src/util/statistics_stats.cpp:57]:
```suggestion
}

```

File: src/util/statistics_stats.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+IntStat& IntStat::operator+=(int64_t val)
+{
+  if constexpr (Configuration::isStatisticsBuild())
+  {
+    d_data->d_value += val;
+  }
+  return *this;
+}
```
Comment (aniemetz) [inline src/util/statistics_stats.cpp:65]:
```suggestion
}

```

File: src/util/statistics_stats.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+void IntStat::maxAssign(int64_t val)
+{
+  if (d_data->d_value < val)
+  {
+    d_data->d_value = val;
+  }
+}
```
Comment (aniemetz) [inline src/util/statistics_stats.cpp:72]:
```suggestion
}

```

File: src/util/statistics_stats.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    PrettyCheckArgument(!d_data->d_running, *this, "timer already running");
+    d_data->d_start = StatisticTimerValue::clock::now();
+    d_data->d_running = true;
+  }
+}
+void TimerStat::stop()
+{
+  if constexpr (Configuration::isStatisticsBuild())
+  {
+    AlwaysAssert(d_data->d_running) << "timer not running";
```
Comment (aniemetz) [inline src/util/statistics_stats.cpp:94]: This doesn't have to be an `AlwaysAssert`. Rather, it probably shouldn't.

File: src/util/statistics_stats.h [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+  /** Allow access to private constructor */
+  friend class StatisticsRegistry;
+  /** Value stored for this statistic */
+  using stat_type = StatisticAverageValue;
+  /** Add the value `v` to the running average */
+  AverageStat& operator<<(double v);
+
+ private:
+  AverageStat(stat_type* data) : d_data(data) {}
+  stat_type* d_data;
```
Comment (aniemetz) [inline src/util/statistics_stats.h:64]: Document

File: src/util/statistics_stats.h [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    if constexpr (Configuration::isStatisticsBuild())
+    {
+      d_data->add(val);
+    }
+    return *this;
+  }
+
+ private:
+  HistogramStat(stat_type* data) : d_data(data) {}
+  stat_type* d_data;
```
Comment (aniemetz) [inline src/util/statistics_stats.h:95]: Document

File: src/util/statistics_stats.h [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+  {
+    if constexpr (Configuration::isStatisticsBuild())
+    {
+      d_data->commit();
+    }
+  }
+
+ private:
+  SizeStat(stat_type* data) : d_data(data) {}
+  stat_type* d_data;
```
Comment (aniemetz) [inline src/util/statistics_stats.h:175]: Document

File: src/util/statistics_stats.h [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+  /** Start the timer. Assumes it is not already running. */
+  void start();
+  /** Stop the timer. Assumes it is running. */
+  void stop();
+  /** Checks whether the timer is running. */
+  bool running() const;
+
+ private:
+  TimerStat(stat_type* data) : d_data(data) {}
+  stat_type* d_data;
```
Comment (aniemetz) [inline src/util/statistics_stats.h:209]: Document

File: src/util/statistics_stats.h [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+  /** Disallow assignment */
+  CodeTimer& operator=(const CodeTimer& timer) = delete;
+  /** Start the timer */
+  CodeTimer(TimerStat& timer, bool allow_reentrant = false);
+  /** Stop the timer */
+  ~CodeTimer();
+
+ private:
+  TimerStat& d_timer;
+  bool d_reentrant;
```
Comment (aniemetz) [inline src/util/statistics_stats.h:236]: Document

File: src/util/statistics_stats.h [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+ * this to the first (or outermost) `CodeTimer`.
+ */
+class CodeTimer
+{
+ public:
+  /** Disallow copying */
+  CodeTimer(const CodeTimer& timer) = delete;
+  /** Disallow assignment */
+  CodeTimer& operator=(const CodeTimer& timer) = delete;
+  /** Start the timer */
```
Comment (aniemetz) [inline src/util/statistics_stats.h:229]: Document `allow_reentrant`.

File: src/util/statistics_stats.h [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+  T get() const
+  {
+    if constexpr (Configuration::isStatisticsBuild())
+    {
+      return d_data->d_value;
+    }
+    else
+    {
+      return T();
+    }
```
Comment (aniemetz) [inline src/util/statistics_stats.h:282]:
```suggestion
    return T();
```

File: src/util/statistics_value.cpp [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+std::ostream& operator<<(std::ostream& out, const StatisticBaseValue& sbv)
+{
+  return detail::print(out, sbv.hasValue() ? sbv.getViewer() : StatExportData{});
+}
+
+StatExportData StatisticAverageValue::getViewer() const { return get(); }
+bool StatisticAverageValue::hasValue() const { return d_count > 0; }
+void StatisticAverageValue::print(std::ostream& out) const { out << get(); }
+void StatisticAverageValue::printSafe(int fd) const
```
Comment (aniemetz) [inline src/util/statistics_value.cpp:78]: Add lines between functions (below, too).

File: src/util/statistics_value.h [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+ */
+struct StatisticBaseValue
+{
+  virtual ~StatisticBaseValue();
+  virtual bool hasValue() const = 0;
+  virtual StatExportData getViewer() const = 0;
+  virtual void print(std::ostream&) const = 0;
+  virtual void printSafe(int fd) const = 0;
+
+  bool d_expert = true;
```
Comment (aniemetz) [inline src/util/statistics_value.h:72]: Move documentation from above to down here and document each.

File: src/util/statistics_value.h [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+    if (d_committed)
+    {
+      if constexpr (std::is_integral_v<T>)
+      {
+        return static_cast<int64_t>(*d_committed);
+      }
+      else
+      {
+        return *d_committed;
+      }
```
Comment (aniemetz) [inline src/util/statistics_value.h:242]:
```suggestion
      return *d_committed;
```

File: src/util/statistics_value.h [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+    else if (d_value != nullptr)
+    {
+      if constexpr (std::is_integral_v<T>)
+      {
+        return static_cast<int64_t>(*d_value);
+      }
+      else
+      {
+        return *d_value;
+      }
```
Comment (aniemetz) [inline src/util/statistics_value.h:253]:
```suggestion
      return *d_value;
```

File: src/util/statistics_value.h [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+  /** Prints seconds in fixed-point format */
+  void print(std::ostream& out) const override;
+  /** Prints seconds in fixed-point format */
+  void printSafe(int fd) const override;
+  /** Make sure that we include the time of a currently running timer */
+  duration get() const;
+
+  duration d_duration;
+  time_point d_start;
+  bool d_running;
```
Comment (aniemetz) [inline src/util/statistics_value.h:374]: Document

**Review by aniemetz — CHANGES_REQUESTED** [review]

**Review by aniemetz — APPROVED** [review]

**Pull request merged**